libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,210 @@
+		add(SpecialNode::ExternalDispatch, _contract.receiveFunction());
+
+	m_contract = nullptr;
+	solAssert(!nodeSet(m_currentNode), "Current node not properly reset.");
+
+	return m_graph;
+}
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		solAssert(*_identifier.annotation().requiredLookup == VirtualLookup::Virtual, "");
+
+		auto funType = dynamic_cast<FunctionType const*>(_identifier.annotation().type);
+
+		// For events kind() == Event, so we have an extra check here
+		if (funType && funType->kind() == FunctionType::Kind::Internal)
+		{
+			processFunction(callable->resolveVirtual(*m_contract), _identifier.annotation());
+
+			solAssert(nodeSet(m_currentNode), "");
+		}
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+		m_graph->createdContracts.emplace(&contractType->contractDefinition());
+
+	return true;
+}
+
+void FunctionCallGraphBuilder::endVisit(MemberAccess const& _memberAccess)
+{
+	auto functionType = dynamic_cast<FunctionType const*>(_memberAccess.annotation().type);
+	auto functionDef = dynamic_cast<FunctionDefinition const*>(_memberAccess.annotation().referencedDeclaration);
+	if (!functionType || !functionDef || functionType->kind() != FunctionType::Kind::Internal)
+		return;
+
+	// Super functions
+	if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+		if (ContractType const* type = dynamic_cast<ContractType const*>(_memberAccess.expression().annotation().type))
+		{
+			solAssert(type->isSuper(), "");
+			functionDef = &functionDef->resolveVirtual(*m_contract, type->contractDefinition().superContract(*m_contract));
+		}
+
+	processFunction(*functionDef, _memberAccess.annotation());
+	return;
+}
+
+void FunctionCallGraphBuilder::endVisit(FunctionCall const& _functionCall)
+{
+	auto* functionType = dynamic_cast<FunctionType const*>(_functionCall.expression().annotation().type);
+
+	if (
+		functionType &&
+		functionType->kind() == FunctionType::Kind::Internal &&
+		!functionType->hasDeclaration()
+	)
+		add(m_currentDispatch, &_functionCall);
+}
+
+void FunctionCallGraphBuilder::visitCallable(CallableDeclaration const* _callable)
+{
+	solAssert(!m_graph->edges.count(_callable), "");
+
+	auto previousNode = m_currentNode;
+	m_currentNode = _callable;
+
+	if (nodeSet(previousNode))
+		add(previousNode, _callable);
+
+	_callable->accept(*this);
+
+	m_currentNode = previousNode;
+}
+
+void FunctionCallGraphBuilder::visitConstructor(
+	ContractDefinition const& _contract,
+	vector<ContractDefinition const*>::const_iterator _start,
+	vector<ContractDefinition const*>::const_iterator _end
+)
+{
+	if (_start != _end)
+		visitConstructor(**_start, _start + 1, _end);
```

Comment:
IMHO passing around these iterators and recursing like this is a really weird thing to do :-). Especially since we usually say we try to avoid using iterators even locally, passing them around as function arguments is quite a huge step :-).
The whole thing is called exactly once from ``FunctionCallGraphBuilder::create``. Why not just loop over ``linearizedBaseContracts`` there and just have this function here visit exactly one constructor and not care about base constructors? That'd be significantly simpler, wouldn't it?

Reply:
In case that's the problem: just use ``boost::adaptors::reversed`` on ``linearizedBaseContracts`` in a range-based loop in ``FunctionCallGraphBuilder::create``, that should give you the exact same visitation order as this construct now.

Reply:
I guess my love for recursion made me blind to the obvious simple solution <3